lib/remap-functions/src/parse_key_value.rs

```diff
@@ -0,0 +1,370 @@
+
+        Ok(Box::new(ParseKeyValueFn {
+            value,
+            key_value_delimiter,
+            field_delimiter,
+        }))
+    }
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct ParseKeyValueFn {
+    value: Box<dyn Expression>,
+    key_value_delimiter: Option<Box<dyn Expression>>,
+    field_delimiter: Option<Box<dyn Expression>>,
+}
+
+impl Expression for ParseKeyValueFn {
+    fn execute(&self, state: &mut state::Program, object: &mut dyn Object) -> Result<Value> {
+        let bytes = self.value.execute(state, object)?.try_bytes()?;
+        let value = String::from_utf8_lossy(&bytes);
+
+        let key_value_delimiter = match &self.key_value_delimiter {
+            Some(s) => String::from_utf8_lossy(&s.execute(state, object)?.try_bytes()?).to_string(),
+            None => "=".to_string(),
+        };
+
+        let field_delimiter = match &self.field_delimiter {
+            Some(s) => String::from_utf8_lossy(&s.execute(state, object)?.try_bytes()?).to_string(),
+            None => " ".to_string(),
+        };
+
+        let (_, values) =
+            parse_line(&value, &key_value_delimiter, &field_delimiter).map_err(|e| e.to_string())?;
+
+        Ok(values.into_iter().collect::<BTreeMap<_, _>>().into())
+    }
+
+    fn type_def(&self, state: &state::Compiler) -> TypeDef {
+        self.value
+            .type_def(state)
+            .merge_optional(
+                self.key_value_delimiter
+                    .as_ref()
+                    .map(|key_value_delimiter| key_value_delimiter.type_def(state)),
+            )
+            .merge_optional(
+                self.field_delimiter
+                    .as_ref()
+                    .map(|field_delimiter| field_delimiter.type_def(state)),
+            )
+            .into_fallible(true)
+            .with_constraint(value::Kind::Map)
+    }
+}
+
+/// Parse the line as a separated list of key value pairs.
+fn parse_line<'a>(
+    input: &'a str,
+    key_value_delimiter: &'a str,
+    field_delimiter: &'a str,
+) -> IResult<&'a str, Vec<(String, Value)>> {
+    separated_list1(
+        parse_field_delimiter(field_delimiter),
+        parse_key_value(key_value_delimiter, field_delimiter),
+    )(input)
+}
+
+/// Parses the field_delimiter between the key/value pairs.
+/// If the field_delimiter is a space, we parse as many as we can,
+/// If it is not a space eat any whitespace before our field_delimiter as well as the field_delimiter.
+/// These lifetimes are actually needed.
+#[allow(clippy::needless_lifetimes)]
+fn parse_field_delimiter<'a>(field_delimiter: &'a str) -> impl Fn(&'a str) -> IResult<&str, &str> {
+    move |input| {
+        if field_delimiter == " " {
+            map(many1(tag(field_delimiter)), |_| " ")(input)
+        } else {
+            preceded(space0, tag(field_delimiter))(input)
+        }
+    }
+}
+
+/// Parse a single `key=value` tuple.
+/// These lifetimes are actually needed.
+#[allow(clippy::needless_lifetimes)]
```

Comment:
Could we resolve this here and in the other places by using `'_` as mentioned on https://rust-lang.github.io/rust-clippy/master/#needless_lifetimes ?

Reply:
Yes, that did work.

```rust
fn parse_value(field_delimiter: &str) -> impl Fn(&str) -> IResult<&str, Value> + '_
```

However, changes I have made to get better error messages means the lifetimes are actually needed now, so Clippy is not complaining anymore.

Reply:
Aha, even better :smile: